petastorm/ngram.py

```diff
@@ -98,13 +99,14 @@ class NGram(object):
     key will be the item.
     """
 
-    def __init__(self, fields, delta_threshold, timestamp_field, timestamp_overlap=True):
+    def __init__(self, fields, delta_threshold, timestamp_field, schema, timestamp_overlap=True):
```

Comment:
The thing is that when we create NGram instance we don't yet have the schema (unless we load it manually from the dataset). This is inconvenient.

E.g.

```python
ngram=NGram(...)
with make_reader(url, schema_fields=ngram) as reader:
   ...
```
Then a user would have to do:
```python
schema=load_schema_from_dataset(url) # don't remember the exact function.
ngram=NGram(...)
with make_reader(url, schema_fields=ngram) as reader:
   ...
```

One way to address this is to defer resolution of regex to later, i.e.  

```python
class NGram(object):
...
   def resolve_regex(schema):
      ....
```

then in `make_reader`, once we have the schema at hand we call `ngram.resolve_regex(schema)`.

What do you think?

That way we also don't have to break compatibility with the existing code.

Reply:
I like the option for make_reader to call resolve_regex. I will change it to that.

Reply:
A follow up, considering that users have to give fields/timestamp_field with a unischema field object.  I might be wrong, but most are probably loading in the schema before creating an ngram.

Reply:
What users would do today, they would take an instance of the schema from the code (and not the dataset). This is not great because they could be out of sync. Maybe we just allow it to be a string?